### PR TITLE
add(consensus): Check consensus branch ids in tx verifier

### LINKED ISSUE #9056: bug: sendrawtransaction discrepancy
### What happened?

 I expected to see this happen: Ywallet hasn't updated its consensus branch ID. When submitting a tx via `sendrawtransaction`, zcashd returns

```
error code: -26
error message:
16: bad-tx-consensus-branch-id-mismatch
```

Instead, this happened: zebra returned a transaction ID. However, later calls to `getrawmempool` did not list that txid. Trying to submit the tx again returns 

```
error code: 0
error message:
any transaction with the same effects will be rejected from the mempool until the next chain tip block
```

Look into it to check if it's just an error handling issue or if there is any bigger issue there.

(TODO: check my logs to see if there are more details)

### What were you doing when the issue happened?

_No response_

### Zebra logs

_No response_

### Zebra Version

_No response_

### Which operating systems does the issue happen on?

- [ ] Linux
- [ ] macOS
- [ ] Windows
- [ ] Other OS

### OS details

_No response_

### Additional information

_No response_

### LINKED ISSUE #9064: Mempool can accept V5 txs with incorrect `nConsensusBranchId`
The mempool currently accepts txs with the `nConsensusBranchId` field set to NU5, even though NU6 is active. This breaks the following consensus rule:

> [**NU5** onward] If `effectiveVersion` ≥ 5, the `nConsensusBranchId` field **MUST** match the consensus branch ID used for SIGHASH transaction hashes, as specified in [ZIP-244](https://zips.z.cash/zip-0244). 

This issue has the following consequences:

1. https://forum.zcashcommunity.com/t/zcashd-deprecation-updates/48965/22
2. https://github.com/ZcashFoundation/zebra/actions/runs/11986841342/job/33443036196?pr=9055

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

Close #9064.

### Context

The tx verifier is not checking if the `nConsensusBranchId` field in V5 txs conforms to the following consensus rule:

> [**NU5** onward] If `effectiveVersion` ≥ 5, the `nConsensusBranchId` field **MUST** match the consensus branch ID used for SIGHASH transaction hashes, as specified in [ZIP-244](https://zips.z.cash/zip-0244). 

We have this check in the block verifier. However, since the mempool uses only the tx verifier, it can currently accept txs with an incorrect `nConsenusBranchId`.

### Specifications & References

- The consensus rules in [§ 7.1.2 Transaction Consensus Rules](https://zips.z.cash/protocol/protocol.pdf#txnconsensus).

## Solution

- Add the check mentioned above to the tx verifier.
- Clean-ups:
  - [Move test-only tx methods out of the default impl](https://github.com/ZcashFoundation/zebra/pull/9063/commits/fad17b6587daaf0359e5e13833b5063c88cb7749)

### Tests

- Add unit tests.
- Adjust existing unit tests.

### PR Author's Checklist

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

### CODE REVIEW

**Comment by conradoplg** [comment]
Just for reference I think the 4.10 rules are about the branch ID used in SIGHASH. What we seem to have missed is from 7.1.2:

> the nConsensusBranchId field MUST match the consensus branch ID used for SIGHASH transaction hashes

which we check for transactions in blocks, but not for transactions in the mempool.

File: zebra-consensus/src/transaction/check.rs [hunk 1]
```diff
@@ -495,3 +495,38 @@ pub fn tx_transparent_coinbase_spends_maturity(
...
+/// #### Notes
+///
+/// 1. At the time of writing this check, the `nConsensusBranchId` field is present only in
+///    [`Transaction::V5`].
+/// 2. When deserializing a transaction, Zebra converts the `nConsensusBranchId` into
+///    [`NetworkUpgrade`] and stores it in [`Transaction::V5::network_upgrade`].
+///
+/// [ZIP-252]: <https://zips.z.cash/zip-0252>
+/// [ZIP-253]: <https://zips.z.cash/zip-0253>
+/// [4.10 SIGHASH Transaction Hashing]: <https://zips.z.cash/protocol/protocol.pdf#sighash>
```
Comment (conradoplg) [inline zebra-consensus/src/transaction/check.rs]: Optional (I think we should change this but we can do that in a separate PR if you prefer). I changed the formatting a bit to match other consensus comments in the codebase.

```suggestion
/// Checks the `nConsensusBranchId` field.
///
/// # Consensus
///
/// ## [7.1.2 Transaction Consensus Rules]
///
/// > [N​U​5 onward] If effectiveVersion ≥ 5, the nConsensusBranchId field MUST match the consensus branch ID
///    used for SIGHASH transaction hashes, as specified in [ZIP-244].
///
/// ## [4.10 SIGHASH Transaction Hashing]
///
/// > 1. [**NU5** only, pre-**NU6**] All transactions **MUST** use the **NU5** consensus branch ID
/// >    `0xF919A198` as defined in [ZIP-252].
/// > 2. [**NU6** only] All transactions **MUST** use the **NU6** consensus branch ID `0xC8E71055`
/// >    as defined in [ZIP-253].
///
/// ### Notes
///
/// 1. At the time of writing this check, the `nConsensusBranchId` field is present only in
///    [`Transaction::V5`].
/// 2. When deserializing a transaction, Zebra converts the `nConsensusBranchId` into
///    [`NetworkUpgrade`] and stores it in [`Transaction::V5::network_upgrade`].
///
/// [ZIP-244]: <https://zips.z.cash/zip-0244>
/// [ZIP-252]: <https://zips.z.cash/zip-0252>
/// [ZIP-253]: <https://zips.z.cash/zip-0253>
/// [4.10 SIGHASH Transaction Hashing]: <https://zips.z.cash/protocol/protocol.pdf#sighash>
/// [7.1.2 Transaction Consensus Rules]: <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
```

**Review by conradoplg — COMMENTED** [review]
This looks good, but since it's a consensus change, I'll leave approval for the next reviewer

**Comment by conradoplg** [comment]
I confirmed that running a `sendrawtransaction` with an outdated-Ywallet-generated tx returns:

```
error code: 0
error message:
transaction did not pass consensus validation
```

File: zebra-consensus/src/transaction/check.rs (hunk 1 above)
Comment (upbqdn) [reply zebra-consensus/src/transaction/check.rs]: I think the new rule requires a different implementation. I'll push that tomorrow.

**Review by daira — APPROVED** [review]
utACK for the non-test code. I did not thoroughly review the tests.

File: zebra-consensus/src/transaction/check.rs [hunk 2]
```diff
@@ -495,3 +495,45 @@ pub fn tx_transparent_coinbase_spends_maturity(
...
 }
+
+/// Checks the `nConsensusBranchId` field.
+///
+/// # Consensus
+///
+/// ## [7.1.2 Transaction Consensus Rules]
+///
+/// > [N​U​5 onward] If effectiveVersion ≥ 5, the nConsensusBranchId field MUST match the consensus branch ID
+///    used for SIGHASH transaction hashes, as specified in [ZIP-244].
```
Comment (oxarbitrage) [inline zebra-consensus/src/transaction/check.rs]:
```suggestion
/// > used for SIGHASH transaction hashes, as specified in [ZIP-244].
```

File: zebra-consensus/src/error.rs [hunk 3]
```diff
@@ -239,6 +239,9 @@ pub enum TransactionError {
     #[error("failed to verify ZIP-317 transaction rules, transaction was not inserted to mempool")]
     #[cfg_attr(any(test, feature = "proptest-impl"), proptest(skip))]
     Zip317(#[from] zebra_chain::transaction::zip317::Error),
+
+    #[error("the transaction uses an incorrect consensus branch id")]
```
Comment (oxarbitrage) [inline zebra-consensus/src/error.rs]:
```suggestion
    #[error("transaction uses an incorrect consensus branch id")]
```

**Review by oxarbitrage — DISMISSED** [review]
looks good, thank you @upbqdn

**Comment by upbqdn** [comment]
> Just for reference I think the 4.10 rules are about the branch ID used in SIGHASH. What we seem to have missed is from 7.1.2:
> 
> > the nConsensusBranchId field MUST match the consensus branch ID used for SIGHASH transaction hashes
> 
> which we check for transactions in blocks, but not for transactions in the mempool.

I agree. To reflect this, I updated the implementation in [b83ee0b](https://github.com/ZcashFoundation/zebra/pull/9063/commits/b83ee0b045196b35a24b8e834be2077d7ba5a073) and [4f37471](https://github.com/ZcashFoundation/zebra/pull/9063/commits/4f37471d3e034a109bb32024cde1ce8a25163c47) and the PR description as well.

**Comment by upbqdn** [comment]
Consensus rules often start with conditions such as _"[NU5 onward]"_. We take those conditions as implicit in the code, but we could implement them by checking what NU is active for the thing being checked. Should we start making those conditions explicit?

**Comment by upbqdn** [comment]
I made the _NU5 onward_ condition explicit in [c698507](https://github.com/ZcashFoundation/zebra/pull/9063/commits/c698507c8154b5b2cb6c32c4f1cff525d7e95b4c).

**Review by oxarbitrage — DISMISSED** [review]

File: zebra-chain/src/transaction.rs [hunk 4]
```diff
@@ -335,6 +335,20 @@ impl Transaction {
...
+    ///
+    /// [§ 7.1 Transaction Encoding and Consensus]: <https://zips.z.cash/protocol/protocol.pdf#txnencoding>
+    pub fn effective_version(&self) -> u32 {
+        if self.is_overwintered() {
+            self.version()
+        } else {
+            std::cmp::min(2, self.version())
+        }
+    }
+
```
Comment (arya2) [inline zebra-chain/src/transaction.rs]: Optional/Nitpick: We could update the docs of the existing `version()` method instead of adding this once since Zebra already uses the V1 or V2 variant when deserializing transactions and always includes the overwintered flag when serializing them.

File: zebra-consensus/src/transaction.rs [hunk 5]
```diff
@@ -380,6 +380,7 @@ where
             // Do quick checks first
             check::has_inputs_and_outputs(&tx)?;
             check::has_enough_orchard_flags(&tx)?;
+            check::consensus_branch_id(&tx, req.height(), &network)?;
```
Comment (arya2) [inline zebra-consensus/src/transaction.rs:383]: Could we remove the check in the block verifier now that it's being checked here so it's easier to rule out as a difference between block transaction validation and mempool transaction validation while debugging? If we remove that, we can also remove the `BlockError::WrongTransactionConsensusBranchId` variant.

The check in the block verifier also isn't using `effective_version()`, though in practice `transaction.network_upgrade()` should always be None if `effective_version() < 5` so it's the same logic.

File: zebra-chain/src/transaction.rs [hunk 6]
```diff
@@ -429,32 +443,6 @@ impl Transaction {
         }
     }
 
-    /// Modify the expiry height of this transaction.
-    ///
-    /// # Panics
-    ///
-    /// - if called on a v1 or v2 transaction
-    #[cfg(any(test, feature = "proptest-impl"))]
-    pub fn expiry_height_mut(&mut self) -> &mut block::Height {
```
Comment (arya2) [inline zebra-chain/src/transaction.rs:438]: Moving these test-only methods is a nice refactor.

Nitpick/Optional: We could defer it to try minimizing merge conflicts with the ZSA/NSM changes, but the conflicts would likely be easy to resolve if you'd like to keep it.

File: zebra-consensus/src/transaction/tests.rs [hunk 7]
```diff
@@ -2577,6 +2563,83 @@ fn v5_with_duplicate_orchard_action() {
...
+            0,
+            Amount::try_from(1).expect("valid amount"),
+        );
+
+        let known_utxos = Arc::new(known_utxos);
+
+        let verify = |tx, nu: NetworkUpgrade| {
+            verifier
+                .clone()
+                .oneshot(Request::Block {
```
Comment (arya2) [inline zebra-consensus/src/transaction/tests.rs:2607]: Optional/Nitpick: Could we also check that the tx verifier returns the same error for `Request::Mempool` calls?

**Review by arya2 — DISMISSED** [review]
This looks really good, thank you for the fix and the related cleanups.

I left a few suggestions for cleanups but there are no blockers.

File: zebra-consensus/src/transaction.rs (hunk 5 above)
Comment (upbqdn) [reply zebra-consensus/src/transaction.rs:383]: I was planning to do this in a next PR.

File: zebra-chain/src/transaction.rs (hunk 4 above)
Comment (upbqdn) [reply zebra-chain/src/transaction.rs]: I wrote this so that the implementation aligns with the logic specified by the protocol (since the consensus rule mentions `effectiveVersion` instead of `version`), but it's superfluous. I'll remove it.

File: zebra-consensus/src/transaction/tests.rs (hunk 7 above)
Comment (upbqdn) [reply zebra-consensus/src/transaction/tests.rs:2607]: The `Request::Mempool` request doesn't take any known UTXOs, so I had trouble composing a TX that would pass the validation. Do you have an idea of how we could construct one?

File: zebra-chain/src/transaction.rs (hunk 4 above)
Comment (upbqdn) [reply zebra-chain/src/transaction.rs]: Removed in [3e7b308](https://github.com/ZcashFoundation/zebra/pull/9063/commits/3e7b308b159d6b9f913a12c855d017a2742b9694).

File: zebra-consensus/src/transaction/tests.rs (hunk 7 above)
Comment (upbqdn) [reply zebra-consensus/src/transaction/tests.rs:2607]: I think I have a solution.

Comment (arya2) [reply zebra-consensus/src/transaction/tests.rs:2607]: We could use the mock state service, though it's not necessary if you'd like to resolve this and merge the PR.

**Review by arya2 — DISMISSED** [review]

File: zebra-consensus/src/transaction/tests.rs (hunk 7 above)
Comment (upbqdn) [reply zebra-consensus/src/transaction/tests.rs:2607]: Done in [e970244](https://github.com/ZcashFoundation/zebra/pull/9063/commits/e97024448042e99bf72c067e280dfff686c2c988).

**Review by arya2 — APPROVED** [review]
